src/kernel/src/fs/uio.rs

```diff
@@ -1,163 +1,169 @@
-                Ok(acc + len)
-            }
-        })?;
+/// Represents a length of [`IoVec`] and [`IoVecMut`].
+#[repr(transparent)]
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct IoLen(usize);
 
-        Ok(Self {
-            vecs,
-            bytes_left,
-            offset,
-        })
+impl IoLen {
+    pub const ZERO: Self = Self(0);
+    pub const MAX: Self = Self(0x7fffffff);
+
+    pub fn from_usize(v: usize) -> Result<Self, IoLenError> {
+        let v = Self(v);
+
+        if v > Self::MAX {
+            Err(IoLenError(()))
+        } else {
+            Ok(v)
+        }
     }
 
-    pub fn from_single_vec(vec: &'a mut IoVec<'a>, offset: i64) -> Self {
-        let bytes_left = vec.len;
+    pub fn get(self) -> usize {
+        self.0
+    }
 
-        Self {
-            vecs: std::slice::from_mut(vec),
-            bytes_left,
-            offset,
+    pub fn checked_add(self, rhs: Self) -> Option<Self> {
+        let r = self.0.checked_add(rhs.0).map(IoLen)?;
+
+        if r > Self::MAX {
+            None
+        } else {
+            Some(r)
         }
     }
 
-    pub fn write_with<E>(
-        &mut self,
-        func: impl Fn(&mut IoVec, i64) -> Result<u64, E>,
-    ) -> Result<(), E> {
-        for vec in self.vecs.iter_mut() {
-            let written = func(vec, self.offset)?;
+    pub fn saturating_add(self, rhs: Self) -> Self {
+        let r = Self(self.0.saturating_add(rhs.0));
 
-            self.offset = self.offset.checked_add_unsigned(written).unwrap();
-            self.bytes_left -= written as usize;
+        if r > Self::MAX {
+            Self::MAX
+        } else {
+            r
         }
+    }
 
-        Ok(())
+    pub fn checked_sub(self, rhs: Self) -> Option<Self> {
+        self.0.checked_sub(rhs.0).map(IoLen)
     }
 }
 
-#[derive(Debug, Error, Errno)]
-pub enum IoVecError {
-    #[error("len exceed the maximum value")]
-    #[errno(EINVAL)]
-    MaxLenExceeded,
+impl TryFrom<SysArg> for IoLen {
+    type Error = IoLenError;
+
+    fn try_from(value: SysArg) -> Result<Self, Self::Error> {
+        Self::from_usize(value.get())
+    }
 }
 
-#[derive(Debug, Error, Errno)]
-pub enum CopyInUioError {
-    #[error("too many iovecs")]
-    #[errno(EINVAL)]
-    TooManyVecs,
+impl PartialEq<usize> for IoLen {
+    fn eq(&self, other: &usize) -> bool {
+        self.0 == *other
+    }
+}
 
-    #[error("the sum of iovec lengths is too large")]
-    #[errno(EINVAL)]
-    MaxLenExceeded,
+impl Display for IoLen {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl From<IoLen> for SysOut {
+    fn from(value: IoLen) -> Self {
+        value.0.into()
+    }
+}
+
+/// Represents an error when [`IoLen`] fails to construct.
+#[derive(Debug)]
+pub struct IoLenError(());
+
+impl Error for IoLenError {}
+
+impl Errno for IoLenError {
+    fn errno(&self) -> NonZeroI32 {
+        EINVAL
+    }
+}
+
+impl Display for IoLenError {
```

Comment:
Can't this be derived with thiserror?  You just put the #[error(...)] above the struct